File: CS/Custom Wizard/WizardDataModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Input;
using DevExpress.DataAccess.ConnectionParameters;
using DevExpress.DataAccess.Sql;
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.POCO;
using WpfReportDesignerDataSourceWizard;

namespace CustomDataSourceWizard {
    public class WizardDataModel : IEditableObject {
        [ServiceProperty(Key = "queryEditorService")]
        protected virtual IDialogService EditQueryDialogService {
            get {
                return null;
            }
        }
        [ServiceProperty(Key = "dsEditorService")]
        protected virtual IDialogService EditDataSourceDialogService {
            get {
                return null;
            }
        }
        public event EventHandler<CustomClosingEventArgs> EndEditing;       
        private WizardDataModel Cache {
            get;
            set;
        }
        private object CurrentModel {
            get {
                return this;
            }
        }
        public virtual DataSourceWrapper SelectedDataSource {
            get;
            set;
        }
        public virtual System.Collections.ObjectModel.ObservableCollection<DataSourceWrapper> DataSources { get; set; }
        [BindableProperty(isBindable:false)]
        public List<ConnectionWrapper> DataConnections { get; set; }
        protected WizardDataModel() {
            InitTestData();
        }
        public static WizardDataModel Create() {
            return ViewModelSource.Create(() => new WizardDataModel());
        }
        protected virtual void OnEndEditing(CustomClosingEventArgs e) {
            EventHandler<CustomClosingEventArgs> handler = EndEditing;
            if(handler != null) {
                handler(this, e);
            }
        }
        public SqlDataSource CreateDataSource() {
            if(SelectedDataSource == null) {
                return null;
            }
            return SelectedDataSource.CreateSqlDataSource();
        }

        #region Data Source Add / Edit / Remove
        public void InitDataSource() {
            DataSourceWrapper dsw = DataSourceWrapper.Create();            
            dsw.BeginEdit();
            bool acceptChanges = OpenDataSourceEditor(dsw);
            if(acceptChanges) {
                dsw.DoEndEdit();
                this.DataSources.Add(dsw);
                this.SelectedDataSource = dsw;
            } else {
                dsw.DoCancelEdit();
            }
        }
        public void EditDataSource() {
            DataSourceWrapper dsw = this.SelectedDataSource;
            dsw.BeginEdit();
            bool acceptChanges = OpenDataSourceEditor(dsw);
            if(acceptChanges) {
               dsw.DoEndEdit();                
            } else {
                dsw.DoCancelEdit();
            }
        }
        private bool OpenDataSourceEditor(DataSourceWrapper dsw) {
            UICommand okCommand = new UICommand() {
                Caption = "OK",
                IsCancel = false,
                IsDefault = true,
                Command = new DelegateCommand<CancelEventArgs>(x => {
                }, x => {
                    return string.IsNullOrEmpty(dsw.Error);
                }),
            };
            UICommand cancelCommand = new UICommand() {
                Id = MessageBoxResult.Cancel,
                Caption = "Cancel",
                IsCancel = true,
                IsDefault = false,
            };
            UICommand result = EditDataSourceDialogService.ShowDialog(
                dialogCommands: new List<UICommand>() { okCommand, cancelCommand },
                title: "Edit Data Source",
                viewModel: new object[] { dsw, this.DataConnections });
            bool acceptChanges = result == okCommand;
            return acceptChanges;
        }
        public bool CanAddDataSource() {
            return true;
        }       
        public bool CanEditDataSource() {
            return this.SelectedDataSource != null;
        }
        public void RemoveSelectedDataSource() {
            this.DataSources.Remove(this.SelectedDataSource);
        }
        public bool CanRemoveSelectedDataSource() {
            return (this.DataSources.Count > 0 && this.SelectedDataSource != null);
        }
        #endregion
        #region Query Add / Edit / Remove
        public void AddQuery() {
            QueryWrapper qw = QueryWrapper.Create();
            qw.OwnerDataSource = this.SelectedDataSource;
            qw.BeginEdit();
            bool acceptChanges = OpenQueryEditor(qw);
            if(acceptChanges) {
                qw.DoEndEdit();
                this.SelectedDataSource.Queries.Add(qw);
                this.SelectedDataSource.SelectedQuery = qw;
            } else {
                qw.DoCancelEdit();
            }
        }
        private bool OpenQueryEditor(QueryWrapper qw) {
            UICommand okCommand = new UICommand() {
                Caption = "OK",
                IsCancel = false,
                IsDefault = true,
                Command = new DelegateCommand<CancelEventArgs>(x => {
                }, x => {
                    return string.IsNullOrEmpty(qw.Error);
                }),
            };
            UICommand cancelCommand = new UICommand() {
                Id = MessageBoxResult.Cancel,
                Caption = "Cancel",
                IsCancel = true,
                IsDefault = false,
            };
            UICommand  result = EditQueryDialogService.ShowDialog(
                dialogCommands: new List<UICommand>() { okCommand, cancelCommand }, 
                title:"Edit Query", 
                viewModel:qw);

            bool acceptChanges = result == okCommand;
            return acceptChanges;
        }
        public bool CanAddQuery() {
            return (this.SelectedDataSource!=null && this.SelectedDataSource.Queries != null);
        }
        public void EditQuery() {
            QueryWrapper qw = this.SelectedDataSource.SelectedQuery;
            qw.BeginEdit();
            bool acceptChanges = OpenQueryEditor(qw);
            if(acceptChanges) {
                qw.DoEndEdit();
                this.SelectedDataSource.SelectedQuery = qw;
            } else {
                qw.DoCancelEdit();
            }
        }
        public bool CanEditQuery() {
            return this.SelectedDataSource != null && this.SelectedDataSource.SelectedQuery != null;
        }
        public void RemoveQuery() {
            this.SelectedDataSource.RemoveSelectedQuery();
        }
        public bool CanRemoveQuery() {
            return (this.SelectedDataSource != null && this.SelectedDataSource.SelectedQuery != null);
        }
        #endregion        
        #region Test data source initialization
        private void InitTestData() {
            this.DataConnections = new List<ConnectionWrapper>();
            this.DataSources = new ObservableCollection<DataSourceWrapper>();

            CreateNorthwind();
            CreateDepartmentsWorks();
            CreateGSP();

            AddInvalidConnection();
        }
        private void CreateGSP() {
            Access97ConnectionParameters connectionParameters = new Access97ConnectionParameters(@"|DataDirectory|\AppData\gsp.mdb", "", "");
            ConnectionWrapper cw = InitConnection("GSPConnection", connectionParameters);
            DataSourceWrapper dsw = DataSourceWrapper.Create();
            dsw.Name = "GSP";
            dsw.Connection = cw;
            this.DataSources.Add(dsw);
        }
        private void CreateDepartmentsWorks() {
            Access97ConnectionParameters connectionParameters = new Access97ConnectionParameters(@"|DataDirectory|\AppData\Departments.mdb", "", "");
            ConnectionWrapper cw = InitConnection("DepartmentsConnection", connectionParameters);
            DataSourceWrapper dsw = DataSourceWrapper.Create();
            dsw.Name = "Departments";
            dsw.Connection = cw;
            QueryWrapper qwSalesPerson = QueryWrapper.Create();
            qwSalesPerson.OwnerDataSource = dsw;
            qwSalesPerson.Name = "Departments";
            qwSalesPerson.Sql = "select * from Departments";
            dsw.Queries.Add(qwSalesPerson);

            this.DataSources.Add(dsw);
        }
        private void CreateNorthwind() {
            DataConnectionParametersBase connectionParameters = new Access97ConnectionParameters(@"|DataDirectory|\AppData\nwind.mdb", "", "");
            ConnectionWrapper cw = InitConnection("NorthwindConnection", connectionParameters);
            DataSourceWrapper dsw = DataSourceWrapper.Create();
            dsw.Name = "Northwind";
            dsw.Connection = cw;
            QueryWrapper qwCategories = QueryWrapper.Create();
            qwCategories.OwnerDataSource = dsw;
            qwCategories.Name = "Categories";
            qwCategories.Sql = "select * from Categories";
            dsw.Queries.Add(qwCategories);
            QueryWrapper qwProducts = QueryWrapper.Create();
            qwProducts.OwnerDataSource = dsw;
            qwProducts.Name = "Products";
            qwProducts.Sql = "select * from Products";
            dsw.Queries.Add(qwProducts);
            this.DataSources.Add(dsw);
        }
        private ConnectionWrapper InitConnection(string connectionName, DataConnectionParametersBase connectionParameters) {
            ConnectionWrapper cw = ConnectionWrapper.Create();
            cw.ConnectionName = connectionName;
            cw.ConnectionParameters = connectionParameters;
            this.DataConnections.Add(cw);
            return cw;
        }
        private void AddInvalidConnection() {
            ConnectionWrapper cw = InitConnection("InvalidConnection", new CustomStringConnectionParameters("invalidConnection"));
        }        
        #endregion
        #region View Model Edit & Cancel methods
        public void DoCancelEdit() {
            this.CancelEdit();
            OnEndEditing(new CustomClosingEventArgs(CloseMode.Cancel));
        }
        public bool CanDoCancelEdit() {
            return true;
        }
        public void DoEndEdit() {
            this.EndEdit();
            OnEndEditing(new CustomClosingEventArgs(CloseMode.Apply));
        }
        public virtual bool CanDoEndEdit() {
            return (this.SelectedDataSource != null && this.SelectedDataSource.SelectedQuery != null);
        }
        #endregion
        #region IEditableObject Members
        [Command(isCommand: false)]
        public void BeginEdit() {
            Cache = ViewModelSource.Create(() => new WizardDataModel());
            foreach(var info in CurrentModel.GetType().GetProperties()) {
                if(!info.CanRead || !info.CanWrite)
                    continue;
                var oldValue = info.GetValue(CurrentModel, null);
                Cache.GetType().GetProperty(info.Name).SetValue(Cache, oldValue, null);
            }
        }
         [Command(isCommand: false)]
        public void CancelEdit() {
            foreach(var info in CurrentModel.GetType().GetProperties()) {
                if(!info.CanRead || !info.CanWrite)
                    continue;
                var oldValue = info.GetValue(Cache, null);
                CurrentModel.GetType().GetProperty(info.Name).SetValue(CurrentModel, oldValue, null);
            }
        }
         [Command(isCommand: false)]
        public void EndEdit() {
            Cache = default(WizardDataModel);
        }
        #endregion
    }
}

File: VB/Custom Wizard/WizardDataModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows
Imports DevExpress.DataAccess.ConnectionParameters
Imports DevExpress.DataAccess.Sql
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO

Namespace CustomDataSourceWizard
    Public Class WizardDataModel
        Implements IEditableObject

        <ServiceProperty(Key := "queryEditorService")> _
        Protected Overridable ReadOnly Property EditQueryDialogService() As IDialogService
            Get
                Return Nothing
            End Get
        End Property
        <ServiceProperty(Key := "dsEditorService")> _
        Protected Overridable ReadOnly Property EditDataSourceDialogService() As IDialogService
            Get
                Return Nothing
            End Get
        End Property
        Public Event EndEditing As EventHandler(Of CustomClosingEventArgs)
        Private Property Cache() As WizardDataModel
        Private ReadOnly Property CurrentModel() As Object
            Get
                Return Me
            End Get
        End Property
        Public Overridable Property SelectedDataSource() As DataSourceWrapper
        Public Overridable Property DataSources() As ObservableCollection(Of DataSourceWrapper)
        <BindableProperty(False)> _
        Public Property DataConnections() As List(Of ConnectionWrapper)
        Protected Sub New()
            InitTestData()
        End Sub
        Public Shared Function Create() As WizardDataModel
            Return ViewModelSource.Create(Function() New WizardDataModel())
        End Function
        Protected Overridable Sub OnEndEditing(ByVal e As CustomClosingEventArgs)
            Dim handler As EventHandler(Of CustomClosingEventArgs) = EndEditingEvent
            If handler IsNot Nothing Then
                handler(Me, e)
            End If
        End Sub
        Public Function CreateDataSource() As SqlDataSource
            If SelectedDataSource Is Nothing Then
                Return Nothing
            End If
            Return SelectedDataSource.CreateSqlDataSource()
        End Function

        #Region "Data Source Add / Edit / Remove"
        Public Sub InitDataSource()
            Dim dsw As DataSourceWrapper = DataSourceWrapper.Create()
            Dim editableObject As IEditableObject = dsw
            editableObject.BeginEdit()
            Dim acceptChanges As Boolean = OpenDataSourceEditor(dsw)
            If acceptChanges Then
                dsw.DoEndEdit()
                DataSources.Add(dsw)
                SelectedDataSource = dsw
            Else
                dsw.DoCancelEdit()
            End If
        End Sub
        Public Sub EditDataSource()
            Dim dsw As DataSourceWrapper = SelectedDataSource
            Dim editableObject As IEditableObject = dsw
            editableObject.BeginEdit()
            Dim acceptChanges As Boolean = OpenDataSourceEditor(dsw)
            If acceptChanges Then
               dsw.DoEndEdit()
            Else
                dsw.DoCancelEdit()
            End If
        End Sub
        Private Function OpenDataSourceEditor(ByVal dsw As DataSourceWrapper) As Boolean
            Dim dataErrorInfo As IDataErrorInfo = dsw
            Dim okCommand As UICommand = New UICommand With {.Caption = "OK", .IsCancel = False, .IsDefault = True, .Command = New DelegateCommand(Of CancelEventArgs)(Sub(x As CancelEventArgs)
                End Sub, Function(x As CancelEventArgs) String.IsNullOrEmpty(dataErrorInfo.Error))}
            Dim cancelCommand As UICommand = New UICommand With {.Id = MessageBoxResult.Cancel, .Caption = "Cancel", .IsCancel = True, .IsDefault = False}
            Dim result As UICommand = EditDataSourceDialogService.ShowDialog(dialogCommands:= New List(Of UICommand) From {okCommand, cancelCommand}, title:= "Edit Data Source", viewModel:= New Object() { dsw, DataConnections })
            Dim acceptChanges As Boolean = result Is okCommand
            Return acceptChanges
        End Function
        Public Function CanAddDataSource() As Boolean
            Return True
        End Function
        Public Function CanEditDataSource() As Boolean
            Return SelectedDataSource IsNot Nothing
        End Function
        Public Sub RemoveSelectedDataSource()
            DataSources.Remove(SelectedDataSource)
        End Sub
        Public Function CanRemoveSelectedDataSource() As Boolean
            Return DataSources.Count > 0 AndAlso SelectedDataSource IsNot Nothing
        End Function
        #End Region
        #Region "Query Add / Edit / Remove"
        Public Sub AddQuery()
            Dim qw = QueryWrapper.Create()
            Dim editableObject As IEditableObject = qw
            qw.OwnerDataSource = SelectedDataSource
            editableObject.BeginEdit()
            Dim acceptChanges As Boolean = OpenQueryEditor(qw)
            If acceptChanges Then
                qw.DoEndEdit()
                SelectedDataSource.Queries.Add(qw)
                SelectedDataSource.SelectedQuery = qw
            Else
                qw.DoCancelEdit()
            End If
        End Sub
        Private Function OpenQueryEditor(ByVal qw As QueryWrapper) As Boolean
            Dim okCommand = New UICommand With {.Caption = "OK", .IsCancel = False, .IsDefault = True, .Command = New DelegateCommand(Of CancelEventArgs)(Sub(x As CancelEventArgs)
                End Sub, Function(x As CancelEventArgs) String.IsNullOrEmpty(DirectCast(qw, IDataErrorInfo).Error))}
            Dim cancelCommand As New UICommand() With {.Id = MessageBoxResult.Cancel, .Caption = "Cancel", .IsCancel = True, .IsDefault = False}
            Dim result As UICommand = EditQueryDialogService.ShowDialog(dialogCommands:= New List(Of UICommand) From {okCommand, cancelCommand}, title:= "Edit Query", viewModel:= qw)

            Dim acceptChanges As Boolean = result Is okCommand
            Return acceptChanges
        End Function
        Public Function CanAddQuery() As Boolean
            Return SelectedDataSource IsNot Nothing AndAlso SelectedDataSource.Queries IsNot Nothing
        End Function
        Public Sub EditQuery()
            Dim qw As QueryWrapper = SelectedDataSource.SelectedQuery
            Dim editableObject As IEditableObject = qw
            editableObject.BeginEdit()
            Dim acceptChanges As Boolean = OpenQueryEditor(qw)
            If acceptChanges Then
                qw.DoEndEdit()
                SelectedDataSource.SelectedQuery = qw
            Else
                qw.DoCancelEdit()
            End If
        End Sub
        Public Function CanEditQuery() As Boolean
            Return SelectedDataSource IsNot Nothing AndAlso SelectedDataSource.SelectedQuery IsNot Nothing
        End Function
        Public Sub RemoveQuery()
            Me.SelectedDataSource.RemoveSelectedQuery()
        End Sub
        Public Function CanRemoveQuery() As Boolean
            Return SelectedDataSource IsNot Nothing AndAlso SelectedDataSource.SelectedQuery IsNot Nothing
        End Function
        #End Region
        #Region "Test data source initialization"
        Private Sub InitTestData()
            Me.DataConnections = New List(Of ConnectionWrapper)()
            Me.DataSources = New ObservableCollection(Of DataSourceWrapper)()

            CreateNorthwind()
            CreateDepartmentsWorks()
            CreateGSP()

            AddInvalidConnection()
        End Sub
        Private Sub CreateGSP()
            Dim connectionParameters As New Access97ConnectionParameters("|DataDirectory|\AppData\gsp.mdb", "", "")
            Dim cw As ConnectionWrapper = InitConnection("GSPConnection", connectionParameters)
            Dim dsw As DataSourceWrapper = DataSourceWrapper.Create()
            dsw.Name = "GSP"
            dsw.Connection = cw
            Me.DataSources.Add(dsw)
        End Sub
        Private Sub CreateDepartmentsWorks()
            Dim connectionParameters As New Access97ConnectionParameters("|DataDirectory|\AppData\Departments.mdb", "", "")
            Dim cw As ConnectionWrapper = InitConnection("DepartmentsConnection", connectionParameters)
            Dim dsw As DataSourceWrapper = DataSourceWrapper.Create()
            dsw.Name = "Departments"
            dsw.Connection = cw
            Dim qwSalesPerson As QueryWrapper = QueryWrapper.Create()
            qwSalesPerson.OwnerDataSource = dsw
            qwSalesPerson.Name = "Departments"
            qwSalesPerson.Sql = "select * from Departments"
            dsw.Queries.Add(qwSalesPerson)

            Me.DataSources.Add(dsw)
        End Sub
        Private Sub CreateNorthwind()
            Dim connectionParameters As DataConnectionParametersBase = New Access97ConnectionParameters("|DataDirectory|\AppData\nwind.mdb", "", "")
            Dim cw As ConnectionWrapper = InitConnection("NorthwindConnection", connectionParameters)
            Dim dsw As DataSourceWrapper = DataSourceWrapper.Create()
            dsw.Name = "Northwind"
            dsw.Connection = cw
            Dim qwCategories As QueryWrapper = QueryWrapper.Create()
            qwCategories.OwnerDataSource = dsw
            qwCategories.Name = "Categories"
            qwCategories.Sql = "select * from Categories"
            dsw.Queries.Add(qwCategories)
            Dim qwProducts As QueryWrapper = QueryWrapper.Create()
            qwProducts.OwnerDataSource = dsw
            qwProducts.Name = "Products"
            qwProducts.Sql = "select * from Products"
            dsw.Queries.Add(qwProducts)
            Me.DataSources.Add(dsw)
        End Sub
        Private Function InitConnection(ByVal connectionName As String, ByVal connectionParameters As DataConnectionParametersBase) As ConnectionWrapper
            Dim cw As ConnectionWrapper = ConnectionWrapper.Create()
            cw.ConnectionName = connectionName
            cw.ConnectionParameters = connectionParameters
            Me.DataConnections.Add(cw)
            Return cw
        End Function
        Private Sub AddInvalidConnection()
            Dim cw As ConnectionWrapper = InitConnection("InvalidConnection", New CustomStringConnectionParameters("invalidConnection"))
        End Sub
        #End Region
        #Region "View Model Edit & Cancel methods"
        Public Sub DoCancelEdit()
            DirectCast(Me, IEditableObject).CancelEdit()
            OnEndEditing(New CustomClosingEventArgs(CloseMode.Cancel))
        End Sub
        Public Function CanDoCancelEdit() As Boolean
            Return True
        End Function
        Public Sub DoEndEdit()
            DirectCast(Me, IEditableObject).EndEdit()
            OnEndEditing(New CustomClosingEventArgs(CloseMode.Apply))
        End Sub
        Public Overridable Function CanDoEndEdit() As Boolean
            Return SelectedDataSource IsNot Nothing AndAlso SelectedDataSource.SelectedQuery IsNot Nothing
        End Function
        #End Region
        #Region "IEditableObject Members"
        Private Sub IEditableObject_BeginEdit() Implements IEditableObject.BeginEdit
            Cache = ViewModelSource.Create(Function() New WizardDataModel())
            For Each info In CurrentModel.GetType().GetProperties()
                If (Not info.CanRead) OrElse (Not info.CanWrite) Then
                    Continue For
                End If
                Dim oldValue = info.GetValue(CurrentModel, Nothing)
                Cache.GetType().GetProperty(info.Name).SetValue(Cache, oldValue, Nothing)
            Next info
        End Sub

        Private Sub IEditableObject_CancelEdit() Implements IEditableObject.CancelEdit
            For Each info In CurrentModel.GetType().GetProperties()
                If (Not info.CanRead) OrElse (Not info.CanWrite) Then
                    Continue For
                End If
                Dim oldValue = info.GetValue(Cache, Nothing)
                CurrentModel.GetType().GetProperty(info.Name).SetValue(CurrentModel, oldValue, Nothing)
            Next info
        End Sub

        Private Sub IEditableObject_EndEdit() Implements IEditableObject.EndEdit
            Cache = Nothing
        End Sub
        #End Region
    End Class
End Namespace

